lnwallet/wallet.go

```diff
@@ -1274,13 +1275,14 @@ func (l *LightningWallet) WithCoinSelectLock(f func() error) error {
 }
 
 // selectCoinsAndChange performs coin selection in order to obtain witness
-// outputs which sum to at least 'numCoins' amount of satoshis. If coin
-// selection is successful/possible, then the selected coins are available
-// within the passed contribution's inputs. If necessary, a change address will
-// also be generated.
+// outputs which sum to at least 'amt' amount of satoshis. If necessary,
+// a change address will also be generated. If coin selection is
+// successful/possible, then the selected coins and change outputs are
+// returned. This method locks the selected outputs, and a function closure to
+// unlock them in case of an error is returned.
 func (l *LightningWallet) selectCoinsAndChange(feeRate SatPerKWeight,
-	amt btcutil.Amount, minConfs int32,
-	contribution *ChannelContribution) error {
+	amt btcutil.Amount, minConfs int32) ([]*wire.TxIn, []*wire.TxOut,
```

Comment:
Would be more readable if the return parameters were lifted into a struct that looks something like: 
```go
type selectedCoins struct {
    Coins []*wire.TxIn
    Change []*wire.TxOut
    ReleaseCoins func() 
}
```

Reply:
Good idea! Done: https://github.com/lightningnetwork/lnd/pull/3165/commits/b136a27562cfa91d569f8a7588d1c42dbaca09d6